### DISCUSSION TITLE
Force Unmount `Script` component from the DOM

### DISCUSSION THREAD

**Discussion by snelsi** [vercel/next.js#42061] [Help]
I'm using `next/script` component to render some inline page analytics that needs to trigger every time the particular page is rendered. However, it looks like this script tag isn't removed from the DOM once the actual `Script` component is unmounted. As a result, when navigating the site, different scripts accumulate from previously visited pages and don't want to unmount or execute a second time when revisiting the page.

```tsx
/index.js
const HomePage = () => (
  <Script
    id="home"
    key="home"
    dangerouslySetInnerHTML={{ __html: homePageInlineAnalytics }}
  />
)
```

```tsx
/about.js
const AboutPage = () => (
  <Script
    id="about"
    key="about"
    dangerouslySetInnerHTML={{ __html: aboutPageInlineAnalytics }}
  />
)
```

Checkout this sandbox:
https://codesandbox.io/s/polished-sky-njelec?file=/pages/index.js

![image](https://user-images.githubusercontent.com/15171036/198651793-6590d9f8-2c04-415e-bd93-f62c3ae0bcce.png)

Is there a way/prop to force the script to unmount from the DOM once the `next/script` is unmounted?

**Comment by snelsi** [comment]
Oh and yeah, I can't move those scripts to `useEffect` for technical reasons. The actual content of the scripts is edited & fetched from the CMS.

**Comment by icyJoseph** [comment]
Mm an ugly hack... if the id is `opaque` for you:

```tsx
import type { NextPage } from "next";
import Link from "next/link";
import Router from "next/router";
import Script from "next/script";
import { useEffect, useMemo } from "react";

let key = 0;
const getKey = () => key++; // strictly increasing or ++key, whichever

const Blog: NextPage = () => {
  const id = useMemo(getKey, []);

  useEffect(() => {
  // clean up on navigation 
    const handler = () => {
      const script = document.getElementById("blog-" + id);
      script?.parentElement?.removeChild(script);
    };

    Router.events.on("routeChangeStart", handler);
    return () => {
      Router.events.off("routeChangeStart", handler);
    };
  }, [id]);

  return (
    <div>
      Blog{" "}
      <Link href="/" className="foo">
        <span>home</span>
      </Link>
      <p>-</p>
      <Link href="/foo">Foo</Link>
      <Script
        id={`blog-${id}`}
        dangerouslySetInnerHTML={{ __html: `console.log("blog")` }}
      />
    </div>
  );
};

export default Blog;

```

It's a bad hack though because the next/script cache thinks new scripts are coming through. At the end of the day that's the big problem, the cache is not being cleaned up, because this component aims to load once, and then keep things available for others to use.
